client/ayon_usd/standalone/usd/pinning/__pinning_file_generation_funcs.py

```diff
@@ -0,0 +1,303 @@
+    uri = re.sub(re.compile(r":SDF_FORMAT_ARGS.*$"), "", ref)
+    return uri
+
+
+def _resolve_udim(udim_identifier: str, layer: Sdf.Layer) -> Dict[str, str]:
+    udim_data = {}
+    resolved_udims = UsdShade.UdimUtils.ResolveUdimTilePaths(
+        str(udim_identifier), layer
+    )
+
+    for path, tile in resolved_udims:
+        udim_replaced_identifier = udim_identifier.replace(
+            "<UDIM>",
+            tile,
+        )
+        udim_data[udim_replaced_identifier] = path
+    return udim_data
+
+
+# TODO refactor so that this function has a gather block and a write block currently all individual blocks write to the identifier_to_path_dict and that makes sanitizing the data hard
+def get_asset_dependencies(
+    layer_path: str, resolver: Ar.Resolver, processed_layers: Optional[List[str]] = None
+) -> Dict[str, str]:
+    """Return mapping from all used asset identifiers to the resolved filepaths.
+
+    Recursively traverse `Sdf.Layer` and get all their asset dependencies.
+    For each asset identifier map it to its resolved filepath.
+
+    Args:
+        layer_path: Usd layer path to be taken as the root layer
+
+    Returns: Mapping from asset identifier to their resolved paths
+
+    """
+    layer_path = _remove_sdf_args(layer_path)
+    if not layer_path:
+        return {}
+    if isinstance(layer_path, Ar.ResolvedPath):
+        layer_path = layer_path.GetPathString()
+
+    identifier_to_path: Dict[str, str] = {}
+
+    resolved_layer_path: Ar.ResolvedPath = resolver.Resolve(layer_path)
+
+    if not processed_layers:
+        processed_layers = [resolved_layer_path.GetPathString()]
+
+    elif resolved_layer_path.GetPathString() in processed_layers:
+        return {}
+
+    else:
+        processed_layers.append(resolved_layer_path.GetPathString())
+
+    layer: Sdf.Layer = Sdf.Layer.FindOrOpen(resolved_layer_path)
+    identifier_to_path[layer_path] = resolved_layer_path.GetPathString()
+    prim_spec_file_paths: List[str] = _get_prim_spec_hierarchy_external_refs(
+        layer.pseudoRoot, layer
+    )
+    for identifier in prim_spec_file_paths:
+        identifier = _remove_sdf_args(identifier)
+        resolved_path = resolver.Resolve(layer.ComputeAbsolutePath(identifier))
+        resolved_path_str = resolved_path.GetPathString()
+        identifier_to_path[identifier] = resolved_path_str
+
+        if "<UDIM>" in resolved_path_str:
+            # Include all tiles/paths of the UDIM
+            udim_data = _resolve_udim(identifier, layer)
+            identifier_to_path.update(udim_data)
+
+    asset_identifier_list: List[str] = layer.GetCompositionAssetDependencies()
+
+    for ref in asset_identifier_list:
+
+        resolved_path = resolver.Resolve(layer.ComputeAbsolutePath(ref))
+
+        ref = _remove_sdf_args(ref)
+        identifier_to_path[ref] = resolved_path.GetPathString()
+
+        recursive_result = get_asset_dependencies(
+            resolved_path.GetPathString(),
```

Comment:
Would it be nicer to re-use the `str`?
```suggestion
        resolved_path = resolver.Resolve(layer.ComputeAbsolutePath(ref))
        resolved_path_str = resolved_path.GetPathString()
        ref = _remove_sdf_args(ref)
        identifier_to_path[ref] = resolved_path_str 

        recursive_result = get_asset_dependencies(
            resolved_path_str ,
```

Reply:
fixed in separate commit